rustls/src/crypto/aws_lc_rs/hpke.rs

```diff
@@ -0,0 +1,774 @@
+        ]
+        .concat();
+
+        let key = labeled_expand(
+            &suite_id,
+            labeled_extract(self.hkdf, &suite_id, Some(&shared_secret.0), b"secret", &[]),
+            b"key",
+            &key_schedule_context,
+            self.aead.key_len() as u16,
+        )
+        .map_err(expand_err)?;
+
+        let base_nonce = labeled_expand(
+            &suite_id,
+            labeled_extract(self.hkdf, &suite_id, Some(&shared_secret.0), b"secret", &[]),
+            b"base_nonce",
+            &key_schedule_context,
+            self.aead.nonce_len() as u16,
+        )
+        .map_err(expand_err)?;
+
+        Ok(KeySchedule {
+            aead: self.aead,
+            key,
+            base_nonce,
+            seq_num: 0,
+        })
+    }
+
+    fn hpke_suite_id(&self) -> Vec<u8> {
+        // "The value of suite_id depends on where the KDF is used; it is assumed implicit from the
+        // implementation and not passed as a parameter. If used inside a KEM algorithm, suite_id MUST
+        // start with "KEM" and identify this KEM algorithm; if used in the remainder of HPKE, it MUST
+        // start with "HPKE" and identify the entire ciphersuite in use."
+
+        const PREFIX: &[u8] = b"HPKE";
+
+        let kem_id = u16::from(self.suite.kem).to_be_bytes();
+        let kdf_id = u16::from(self.suite.sym.kdf_id).to_be_bytes();
+        let aead_id = u16::from(self.suite.sym.aead_id).to_be_bytes();
+
+        [PREFIX, &kem_id, &kdf_id, &aead_id].concat()
+    }
+}
+
+impl Hpke for &Suite {
+    fn seal(
+        &mut self,
+        info: &[u8],
+        aad: &[u8],
+        plaintext: &[u8],
+        pub_key: &HpkePublicKey,
+    ) -> Result<(EncapsulatedSecret, Vec<u8>), Error> {
+        let (encap, mut sealer) = Sealer::new(self, info, pub_key)?;
+        Ok((encap, sealer.seal(aad, plaintext)?))
+    }
+
+    fn setup_sealer(
+        &mut self,
+        info: &[u8],
+        pub_key: &HpkePublicKey,
+    ) -> Result<(EncapsulatedSecret, Box<dyn HpkeSealer + 'static>), Error> {
+        let (encap, sealer) = Sealer::new(self, info, pub_key)?;
+        Ok((encap, Box::new(sealer)))
+    }
+
+    fn open(
+        &mut self,
+        enc: &EncapsulatedSecret,
+        info: &[u8],
+        aad: &[u8],
+        ciphertext: &[u8],
+        secret_key: &HpkePrivateKey,
+    ) -> Result<Vec<u8>, Error> {
+        Opener::new(self, enc, info, secret_key)?.open(aad, ciphertext)
+    }
+
+    fn setup_opener(
+        &mut self,
+        enc: &EncapsulatedSecret,
+        info: &[u8],
+        secret_key: &HpkePrivateKey,
+    ) -> Result<Box<dyn HpkeOpener + 'static>, Error> {
+        Ok(Box::new(Opener::new(self, enc, info, secret_key)?))
+    }
+}
+
+impl Debug for Suite {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        self.suite.fmt(f)
+    }
+}
+
+/// Adapts a [KeySchedule] for the role of a [HpkeSealer].
+struct Sealer {
+    key_schedule: KeySchedule,
```

Comment:
Why don't we store the `SealingKey` in the `Sealer` instead of the `KeySchedule`?

Reply:
> Do you mean the shared secret, KemSharedSecret

Ah, no, on second look I suspect you mean the AEAD key.

Reply:
Reworked.

Reply:
This didn't look like what I had in mind. I thought we might be able to just store the `SealingKey` in the `Sealer`? Looking again, I guess that doesn't work because we have to create a new `SealingKey` based on the `nonce_schedule` for every `seal()` operatiion?

Reply:
> Looking again, I guess that doesn't work because we have to create a new SealingKey based on the nonce_schedule for every seal() operatiion?

That's correct. Should I go back to the original form?

Reply:
Reverted to keeping the AEAD bits in the key schedule.